### PR TITLE
Run Map iterations in parallel up to MaxConcurrency

### PR DESCRIPTION
**Pull request by agrare**
Important to note that while it is possible for multiple iterations to be running in parallel externally (aka docker containers), there is still only a single thread checking for branch status so IMO we don't have any concurrency issues with the Context hash here.

### CODE REVIEW

File: examples/map.asl [hunk 1]
```diff
@@ -27,11 +28,15 @@
...
             "Parameters": {
               "uuid.$": "States.UUID()"
-            },
-            "OutputPath": "$.uuid"
+            }
+          },
+          "Sleep": {
+            "Type": "Task",
+            "Resource": "docker://docker.io/agrare/sleep:latest",
+            "End": true
```
Comment (agrare) [inline examples/map.asl:39]: NOTE this gives us a sub-iteration that runs a task that doesn't complete immediately like a Pass state would so it exercises the Workflow.wait event-handling code for child-tasks

File: spec/workflow/states/map_spec.rb [hunk 2]
```diff
@@ -26,7 +26,7 @@
         "Type"           => "Map",
         "InputPath"      => "$.detail",
         "ItemsPath"      => "$.shipped",
-        "MaxConcurrency" => 0,
```
Comment (agrare) [inline spec/workflow/states/map_spec.rb:29]: I was wondering why my spec was hanging :laughing: 
Add a check to raise a workflow error if MaxConcurrency was 0

**Comment by agrare** [comment]
> Method step_nonblock! has a Cognitive Complexity of 6 (exceeds 5 allowed). Consider refactoring

Okay I got this down to 6 (I don't usually chase cognitive complexity but this method was a little gross and now is improved)

File: lib/floe/workflow/states/map.rb [hunk 3]
```diff
@@ -74,7 +72,7 @@ def finish(context)
         def run_nonblock!(context)
           start(context) unless context.state_started?
 
-          loop while step_nonblock!(context) == 0 && running?(context)
+          step_nonblock!(context) while running?(context)
```
Comment (kbrock) [inline lib/floe/workflow/states/map.rb:75]: is this a universal change you want to make?
(not asking for the change, just trying to understand trajectory)

Comment (agrare) [reply lib/floe/workflow/states/map.rb:75]: Specifically for map this was now essentially doing the same running? check twice (ended? is just !running?) so it was more efficient to do this

**Comment by agrare** [comment]
@kbrock can you take a look?

**Comment by Fryguy** [comment]
Is Map + Max Concurrency different than Parallel?

**Review by kbrock — APPROVED** [review]
This has some great changes.

We did discuss possibly not looping the contexts for so many of the operations and using a hash for the contexts. (for future thought)

**Pull request merged**